---
title: "Exploratory analysis of EBI-Metagenomic portal contingency tables\n enriched with ENA metadata"
output:
  pdf_document:
    fig_crop: no
    fig_height: 10
    fig_width: 13
  number_sections: yes
  toc: yes
  fontsize: 10pt
  html_document: default
  word_document: default
date: Bari, June, 2017 
author: Blaise T.F. Alako
header-includes:
    - \usepackage{wallpaper}

---
\URCornerWallPaper{1.0}{//Users/blaise/Desktop/MetagenomicsCourse/logo2.png}
\LRCornerWallPaper{1.0}{//Users/blaise/Desktop/MetagenomicsCourse/enalogo.png}


***  
__Dataset: American Gut Project__  
The American Gut project is the largest crowdsourced citizen science project to date. Fecal, oral, skin, and other body site samples collected from thousands of participants represent the largest human microbiome cohort in existence. Detailed health and lifestyle and diet data associated with each sample is enabling us to deeply examine associations between the human microbiome and factors such as diet (from vegan to near carnivore and everything in between), season, amount of sleep, and disease states such as IBD, diabetes, or autism spectrum disorder-as well as many other factors not listed here. The American Gut project also encompasses the British Gut and Australian Gut projects, widening the cohort beyond North America. As the project continues to grow, we will be able to identify significant associations that would not be possible with smaller, geographically and health/disease status-limited cohorts. 
_ENA website (ERP012803)_ 

***  

We will explore the data generated by the study above. The data is available at:  
https://www.ebi.ac.uk/metagenomics/projects/ERP012803 

**Loading of R packages**  

```{r}
suppressWarnings(suppressMessages(require(stringr)))      # String format
suppressWarnings(suppressMessages(require(ade4)))         # Multivariate analysis
suppressWarnings(suppressMessages(require(ggplot2)))      # Fancy plotting
suppressWarnings(suppressMessages(require(grid)))         # Has the viewport function
suppressWarnings(suppressMessages(require(dplyr)))        # Data manipulation
suppressWarnings(suppressMessages(require(tidyr)))        # Data manipulation
suppressWarnings(suppressMessages(require(factoextra)))   # Visualize result of
suppressWarnings(suppressMessages(require(FactoMineR)))   # multivariate analysis
suppressWarnings(suppressMessages(require(XML)))
suppressWarnings(suppressMessages(require(xml2)))
suppressWarnings(suppressMessages(devtools::install_github("hrbrmstr/xmlview")))
suppressWarnings(suppressMessages(require(xmlview)))
suppressWarnings(suppressMessages(require(xml2)))
suppressWarnings(suppressMessages(require(purrr)))
suppressWarnings(suppressMessages(require(surveillance)))
suppressWarnings(suppressMessages(require('gplots')))
suppressWarnings(suppressMessages(require(webshot)))

```

**Obtain input data for analysis**

```{r}
# Get the analysis summary file
ERP012803 <- "https://www.ebi.ac.uk/metagenomics/projects/ERP012803/download/2.0/export?contentType=text&exportValue=taxonomy_abundances"
ERP012803.meta <- "https://www.ebi.ac.uk/metagenomics/projects/ERP012803/overview/doExport";
gutproject <- tbl_df(read.delim(file=ERP012803))
meta <-  tbl_df(read.csv(file=ERP012803.meta))
meta.MG <- meta %>%   mutate(Sample.Description=gsub(" ","_",Sample.Description),
                         Sample.Description=gsub("American_Gut_Project_","", Sample.Description), 
                         Sample.Description=gsub("American_","", Sample.Description)
                         ) %>% select(Sample.Description, Sample.ID, Run.ID)


```


```{r}
#############################################################
#Extracts extensive information about the sample from the ENA
#############################################################
feature_selection <- c('sample_id','alcohol_frequency','cosmetics_frequency','diet_type',
                       'high_fat_red_meat_frequency','level_of_education','probiotic_frequency',
                       'red_meat_frequency','sex','teethbrushing_frequency','liver_disease',
                       'lung_disease','thyroid','alcohol_types_red_wine','alcohol_types_white_wine',
                       'cardiovascular_disease','fruit_frequency','meat_eggs_frequency',
                       'multivitamin','pregnant','alcohol_consumption','alcohol_types_sour_beers',
                       'diabetes','bmi_cat','cat','kidney_disease','nail_biter','tonsils_removed')

sub_feature <- c('sample_id','alcohol_frequency', 'diet_type','sex','fruit_frequency',
                 'meat_eggs_frequency','multivitamin','alcohol_consumption','bmi_cat','diabetes',
                 'liver_disease','lung_disease','thyroid','cardiovascular_disease',
                 'pregnant','kidney_disease'
                 )

extractSampleInfo <- function(x){
  tag <- xml_text(xml_find_all(x, xpath='.//TAG'))
  tag <- as.character(tag)
  value <- xml_text(xml_find_all(x, xpath='.//VALUE'))
  value <- as.character(value)
  id <- xml_text(xml_find_all(x, xpath='.//PRIMARY_ID'))
  id <- as.character(id)
  sampleInfo <- tbl_df(data.frame(id=id, tag=tag,value=value))
  #Select a subset of features
  sampleInfo <- sampleInfo %>% filter(tag %in% feature_selection) #%>% filter(!value=='Unknown')
  return(sampleInfo)
}

constructQuery <- function( accs=accs){
  info <- paste("Processing ... ", length(accs), ' BiosampleID' , sep="")
  print(info)
  accs <- paste(accs,collapse = ",")
  url <- paste("https://www.ebi.ac.uk/ena/data/view/", accs, "&display=xml", sep="")
  return(url)
}

getXML <- function(url=url){
  xml <- read_xml(url)
  return(xml)
}

retrieveXmlContent <- function(accs=accs){
  query <- constructQuery(accs=accs)
  xmlContent <- getXML(url=query)
  return(xmlContent)
}
# What information are capture for each biosample

biosample_id <- as.character(meta$Sample.ID[1])
biosample.xml <- retrieveXmlContent(accs=biosample_id)

```

__Uncomment the following code to view a sample ENA SAMPLE XML file__

```{r}
#xml_view(biosample.xml, add_filter = TRUE)
#webshot(xml_view(biosample.xml, add_filter = TRUE), "r.png")
# NB: look for TAG that contains alcohol: (.//TAG[contains(.,'alcohol')]) in the XPath box
# How many feature are capture for this sample?
xml_find_all(biosample.xml, './/TAG', ns=xml2::xml_ns(biosample.xml))

#There are 432 feature captured, we should select a subset for downstream analysis

```

**Uncomment the following to retrieve metadata from ENA**

```{r}
#-------------------------------------------------------------
#------ The following is time consuming hence commented. 
#------ The retrieved data is provided as an R object below
#-------------------------------------------------------------
# Since we have > 8000 sample to fetch extensive information, 
# lets us process this in batch of 500

# accs <- split(meta$Sample.ID, ceiling(seq_along(meta$Sample.ID)/500))
# system.time(results<- lapply(accs, retrieveXmlContent))
# 
# # Loop through the chunck in order to extract Sample Information
# # #################################################################
# 
# processEachBiosampleXML <- function(mydata=mydata){
#          print(paste("Processing XML object with .. ",
#                      xml_length(mydata) , " Biosample ID" ,sep=" "))
#           mydata %>% xml_find_all("//SAMPLE") %>%   # Proceed per Biosample ID
#           map (~ extractSampleInfo(.x))       # Extract information per BiosampleID
# 
# }

```

```{r}
# #lapply(results, processEachBiosampleXML)       # Extract information per BiosampleID
# system.time(biosample_content <- plapply(results, processEachBiosampleXML, .parallel=6, .verbose = TRUE))
# 
# #Convert the list into dataframe
# list2df <- function(x){
#    mdf <- tbl_df(do.call(rbind,x))
#    return(mdf)
# }
# 
# #Collapse all sub dataframes in one big table
# system.time(sample.info.list <- plapply(biosample_content, list2df, .parallel=6, .verbose = TRUE))
# sample.info <- sample.info.list %>%
#                  bind_rows() %>% unique() 
#%>% filter(tag %in% feature_selection) %>% filter(!value=='Unknown')
# 
# save(file="gutproject.sample.info.RData", sample.info)

```


```{r}
setwd('/Users/blaise/Desktop/MetagenomicsCourse')
load(file = "gutproject.sample.info.RData")
```

**Preprocess the data**

```{r}
# Retain only lineages with full species name
gutproject <- gutproject %>% 
              rename(taxonomy=X.SampleID) %>%
                #mutate(taxonomy=gsub(".*?s__","",taxonomy)) %>% 
                mutate(taxonomy=gsub(".*?g__","",taxonomy)) %>%
                mutate(taxonomy=gsub(";s__","_", taxonomy)) %>%
                  filter(taxonomy !="" | taxonomy=='_' ) %>%
                    filter(!grepl('_$', taxonomy))
```


```{r}
# Merge counts of species in the same experimental sample
gutproject <- gutproject %>% 
                group_by(taxonomy) %>%
                  summarize_each(funs(sum)) %>%
                    ungroup()
```

```{r}
# Use the species name as the rowname, this replaces the default numerical name 
rownames(gutproject) <- gutproject$taxonomy
```

```{r}
# Transform the wide format data into the long format for the purpose of appending 
# descriptive information
gutproject.long  <- gutproject %>% 
                      gather(Run.ID, freq, ERR1072624:ERR1160857) %>%
                        filter(taxonomy !="Root") 
```


```{r}
# Retain only the sample description and the Run id from the metadata
# Merge ENA metadata with MG metadata.
meta.MG <- meta.MG %>% mutate(Sample.ID=as.character(Sample.ID))
meta.ENA <- sample.info

meta.ENA.MG <- inner_join(meta.ENA, meta.MG, by=c("id"="Sample.ID"))
# select subset of column for downstream analysis
# Remove qualifier with unknown values

ena.meta <- meta.ENA.MG %>% select(Run.ID, tag,value) %>% 
            unique() %>% filter(tag %in% sub_feature) %>%
            mutate(Run.ID=as.character(Run.ID)) %>%
            mutate(Sample.Description=paste(tag,value ,sep=":")) %>%
            select(Run.ID,Sample.Description) %>%  
            filter(!grepl('Unknown', Sample.Description))  %>%
            filter(!grepl('I do not have', Sample.Description))  %>% 
            unique()

mg.meta  <- meta.ENA.MG %>% select(Run.ID, Sample.Description,id) %>% 
  unique() %>% 
  mutate(Run.ID=as.character(Run.ID))

```


```{r}
# What is the data made up of.

meta.strip <- mg.meta %>% select(Sample.Description, Run.ID,id) %>% 
              unique() %>% group_by(Sample.Description) %>%
              mutate(count=n()) %>% ungroup() %>%
              select(Sample.Description, count)  %>% unique()

meta.strip %>% ggplot(aes(x=Sample.Description, y=count)) + 
                geom_bar(stat='identity') + 
                geom_text(aes(label=count))  + theme_bw() +
                      theme(axis.text.x=element_text(angle=45, hjust=1)) + 
                      ggtitle("American gut project") + xlab("Sample source")

```

```{r}

####################################################
# Merge metagenomics contingency table with the 
# the detailed ENA metadata for each biosample_ID
####################################################
system.time(gutproject.ena <- inner_join(gutproject.long, ena.meta, by=c("Run.ID"="Run.ID")))
system.time(gutproject.mg <- inner_join(gutproject.long, mg.meta, by=c("Run.ID"="Run.ID")))

```

__Merge both annotations___

```{r}

system.time(gutproject.ena <- gutproject.ena %>% 
              group_by(taxonomy,Sample.Description) %>%
              mutate(freq=sum(freq)) %>%
              ungroup() %>% 
              select(taxonomy,Sample.Description, freq) %>%
              unique())

system.time(gutproject.mg <- gutproject.mg %>% select(-Run.ID, -id) %>% 
              group_by(taxonomy,Sample.Description) %>% 
              mutate(freq=sum(freq)) %>% 
              ungroup() %>%
              select(taxonomy,Sample.Description, freq) %>% 
              unique())


gutproject.df <- bind_rows(gutproject.ena, gutproject.mg)

```
--Transform the long format to wide format for the purpose of Mutivariate analysis__

```{r}


gutproject.df <- gutproject.df %>%
                     spread(Sample.Description, freq, fill=0) %>%
                       as.data.frame()
gutproject.mat <- gutproject.df %>% select(-c(1)) %>% as.matrix()
rownames(gutproject.mat) <- gsub("\\[|\\]","", perl=TRUE, gutproject.df$taxonomy)
```

```{r}

gutproject.mat <-  gutproject.mat[!rownames(gutproject.mat)=="",]
#Remove rows that sums to 0
gutproject.mat <- gutproject.mat[which(rowSums(gutproject.mat)!=0),]
#Remove columns that sums to 0
gutproject.mat <- gutproject.mat[,which(colSums(gutproject.mat)!=0)]

```

_Is there any association between the species and the data source?_

```{r}
# Perform a chi-square test of independence, this evaluates whether there is a significant
# dependence between species and biological samples

chisq <- chisq.test(gutproject.mat)
chisq
# The species and biological samples are statistically significantly associated (p-value=0)
```

_Perform Corresspondance Anlysis, CA (a generalized from of Principal Component Analysis for categorical data)_

```{r fig.pos="H"}
# We use CA to reduce the dimension of the data without loosing the most important information.
# CA is used to graphically visualize rows points and column points in a low dimensional space

gutproject.ca <- CA(gutproject.mat, graph=FALSE)
```

_Explore the content of the CA output_

```{r}
summary(gutproject.ca, nb.dec = 2, nbelements = 4, ncp= 3)

# The result of the function summary() contains the chi-square statistics and 3 tables:
# Table 1- Eigenvalues, displays the variances and the percentage of vaiances retained by each 
#         dimension.
# Table 2, Contains the coordinates, the contribution and the cos2 (quality of representation 
#          [0-1] of the first 4 active rows variable on the dimension 1, 2 and 3
# Table 3: displays the coordinates, the contribution and the cos2 of the first 4 active column 
#          variables on the dimension 1, 2 and 3
#     
```

_Interprete the CA result above, hint:correlation coef., chi-square statistic_

```{r}
# Correlation coefficient
eig <- get_eigenvalue(gutproject.ca)
# Eigen values is the amount of information retained by each axis.
trace <- sum(eig$eigenvalue)
cor.coef <- sqrt(trace)
cor.coef
# As a rule of thumb a correlation above 0.2 can be considered important 
# (Bendixen 1995, 567; Healey 2013, 289-290)
# The correlation coef. is 0.50in the American gut project dataset, 
# indicating a strong association between row (species) 
# and column(biological samples variables.). 
# A more rigorous approach for examining the association is to use Chi-square statistics. 
# The association is highly significant (p-value=0) 
```

```{r}
# Explore the percentages of inertia explained by the CA dimensions, the scree plot
fviz_screeplot(gutproject.ca, barfill='white', addlabels=TRUE) + theme_bw()
# The point at which the scree plot shows a bend mignt indicate an optimal dimensionality.
```

_Hierarchical Clustering of principal components_

```{r fig.pos="H"}
# Compute hierarchical clustering of species of CA results
gutproject.hcpc <- HCPC(gutproject.ca, graph = TRUE , nb.clust=10, order=TRUE )
```

_Visualize the CA results_

```{r fig.pos="H"}
# Symmetric biplot, whereby both rows (blue) and columns (red) are represented
# in the same space using the principal coordinates. Only the distance between row
# points or the distance between column points can be reliably interpreted. 
# With a symmetric plot, only general statements can be drawn about the pattern. 
# The inter-distance between rows and column can not be interpreted.
fviz_ca_biplot(gutproject.ca, repel=TRUE, select.row = list(contrib = 50), 
               select.col = list(contrib = 50), geom="text") +
  theme_minimal()
```

_Remove labels and add cluster centers_

```{r fig.pos="H"}
# A 3D map of the hierarchical clustering of the principal component.
plot(gutproject.hcpc, choice ="3D.map", draw.tree = FALSE,
     ind.names = FALSE, centers.plot = TRUE, 
     title='Hierarchical clustering of the Principal Components')
```

_Hierarchical Clustering of principal component._

```{r fig.pos="H"}
# Compute hierarchical clustering of biological samples of CA results
res.hcpc <- HCPC(gutproject.ca, cluster.CA = "columns",
                 graph = TRUE , nb.clust=8, order=TRUE )
```

```{r fig.pos="H"}
# Biological sample cluster in the dataset
fviz_cluster(res.hcpc ) + theme_bw()
```

```{r}
# Clustering of sample qualifiers....s
fviz_dend(res.hcpc, horiz=T, rec_fill=TRUE, rect = TRUE, 
          type="rectangle", k=5, sub=NULL, main="PC clustering") + 
     theme( axis.line.x  = element_line(colour = "grey", size = .3, linetype = "solid"))

```

```{r fig.pos="H"}
# transpose the row and column in the gut data and perform
# a new correspondance analysis
#gutproject.ca <- CA(t(gutproject.mat), ncp=4)
```

```{r fig.pos="H"}
# Symmetric biplot 
fviz_ca_biplot(gutproject.ca, repel=TRUE, select.row = list(contrib = 12), 
               select.col = list(contrib = 95), geom="text") +
  theme_minimal()
```

_Remove labels and add cluster centers_

```{r fig.pos="H"}
# 3d hierarchical clustering of the principal components.
plot(res.hcpc, choice ="3D.map", draw.tree = FALSE,
     ind.names = FALSE, centers.plot = TRUE,
     title='Hierarchical clustering of the Principal Components')
```


```{r}
z_score <- function (X){
  pop_sd <- sd(X)*sqrt((length(X)-1)/(length(X)))
  pop_mean <- mean(X)
  z_score_val <- (X - pop_mean)/pop_sd
}
#################################################
# Calculate norm of vector for normalization....
#################################################
norm_vec <- function(x) x/sqrt(sum(x^2))

```
__Association between species and sample source and qualifiers__

```{r}

gutproject.mat.ena <- gutproject.mat[,!grepl('sample',colnames(gutproject.mat))]
gutproject.mat.mg <- gutproject.mat[,grepl('sample',colnames(gutproject.mat))]
gutproject.mat.merge <- gutproject.mat

```

_**Association between species and sample source (MG portal metadata)**_

```{r}
gutproject.mat.mg <-  gutproject.mat.mg[!rownames(gutproject.mat.mg)=="",]
#Remove rows that sums to 0
gutproject.mat.mg <- gutproject.mat.mg[which(rowSums(gutproject.mat.mg)!=0),]
#Remove columns that sums to 0
gutproject.mat.mg <- gutproject.mat.mg[,which(colSums(gutproject.mat.mg)!=0)]


gutproject.ca <- CA(gutproject.mat.mg, graph = FALSE) 
gutproject.ca.ctrb  <- fviz_ca_row(gutproject.ca, alpha.row="contrib", 
                                   select.row=list(contrib=50))
bestref.ctrb <- gutproject.ca.ctrb$data[,"name"]
gutproject.mat.sub <- gutproject.mat.mg[rownames(gutproject.mat.mg) %in% bestref.ctrb, ]
gutproject.mat.norm <- apply(gutproject.mat.sub, MARGIN = 2, 
                             FUN = function(X) (norm_vec(X)))



####################################################
#           .....Heatmap MG metadata ...
####################################################
rc <- rainbow(nrow(gutproject.mat.norm), start=0, end=.3)
cc <- rainbow(ncol(gutproject.mat.norm), start=0, end=.3)
heatmap.2(gutproject.mat.norm, col=cm.colors(255), scale="row",
                       RowSideColors=rc, ColSideColors=cc, margin=c(5, 10),
                       xlab="Sample sources", ylab= "",
                       main="",
                       tracecol="steelblue", density="density",
                       srtRow=55, adjRow=c(0, 1), srtCol=10, adjCol=c(1,1)
)
```

_**Association between species and sample qualifiers (ENA portal metadata)**_

```{r}

gutproject.mat.ena <-  gutproject.mat.ena[!rownames(gutproject.mat.ena)=="",]
#Remove rows that sums to 0
gutproject.mat.ena <- gutproject.mat.ena[which(rowSums(gutproject.mat.ena)!=0),]
#Remove columns that sums to 0
gutproject.mat.ena <- gutproject.mat.ena[,which(colSums(gutproject.mat.ena)!=0)]

gutproject.ca <- CA(gutproject.mat.ena, graph = FALSE) 
gutproject.ca.ctrb  <- fviz_ca_row(gutproject.ca, alpha.row="contrib",
                                   select.row=list(contrib=50))
bestref.ctrb <- gutproject.ca.ctrb$data[,"name"]
gutproject.mat.sub <- gutproject.mat.ena[rownames(gutproject.mat.ena) %in% bestref.ctrb, ]
#corpmat.ctrb <- rcorr(t(mydf.mat.ctrb))
gutproject.mat.norm <- apply(gutproject.mat.sub, MARGIN = 2,
                             FUN = function(X) (norm_vec(X)))

####################################################
#             .....Heatmap ENA metadata ...
####################################################
rc <- rainbow(nrow(gutproject.mat.norm), start=0, end=.3)
cc <- rainbow(ncol(gutproject.mat.norm), start=0, end=.3)
heatmap.2(gutproject.mat.norm, col=cm.colors(255), scale="row",
                       RowSideColors=rc, ColSideColors=cc, margin=c(5, 10),
                       xlab="Sample qualifiers", ylab= "",
                       main="",
                       tracecol="steelblue", density="density",
                       srtRow=55, adjRow=c(0, 1), srtCol=10, adjCol=c(1,1)
)
```

_**Association between species and sample qualifers and source (MG + ENA portal metadata)**_

```{r}

gutproject.mat.merge <-  gutproject.mat.merge[!rownames(gutproject.mat.merge)=="",]
#Remove rows that sums to 0
gutproject.mat.merge <- gutproject.mat.merge[which(rowSums(gutproject.mat.merge)!=0),]
#Remove columns that sums to 0
gutproject.mat.merge <- gutproject.mat.merge[,which(colSums(gutproject.mat.merge)!=0)]

gutproject.ca <- CA(gutproject.mat.merge, graph = FALSE) 
gutproject.ca.ctrb  <- fviz_ca_row(gutproject.ca, alpha.row="contrib", 
                                   select.row=list(contrib=50))
bestref.ctrb <- gutproject.ca.ctrb$data[,"name"]
gutproject.mat.sub <- gutproject.mat.ena[rownames(gutproject.mat.ena) %in% bestref.ctrb, ]
#corpmat.ctrb <- rcorr(t(mydf.mat.ctrb))
gutproject.mat.norm <- apply(gutproject.mat.sub, 
                             MARGIN = 2, FUN = function(X) (norm_vec(X)))

####################################################
#             .....Heatmap Merge metadata ...
####################################################
rc <- rainbow(nrow(gutproject.mat.norm), start=0, end=.3)
cc <- rainbow(ncol(gutproject.mat.norm), start=0, end=.3)
heatmap.2(gutproject.mat.norm, col=cm.colors(255), scale="row",
                       RowSideColors=rc, ColSideColors=cc, margin=c(5, 10),
                       xlab="Sample sources and qualifiers", ylab= "",
                       main="",
                       tracecol="steelblue", density="density",
                       srtRow=55, adjRow=c(0, 1), srtCol=10, adjCol=c(1,1)
)
```




_**APPENDIX**_

_To reproduce this hands on session, your R sessionInfo() must be identical to the following:_

```{r}
sessionInfo()
```
